realtek: Fix kernel 5.15 bootloop on rtl838x

hardfalcon opened the pull request:
Fix the bootloop with kernel 5.15 on rtl838x devices that is caused by a regression introduced with 758c88b969639d0e6b684669d2e54dd1be3102f4.

Tested on a Netgear GS108T v3 and a Netgear GS310TP v1.

Fixes: 758c88b969639d0e6b684669d2e54dd1be3102f4 ("realtek: Whitespace and codestyle cleanup")
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon committed:
realtek: return correct error value for phy ops

A behavioural change was introduced with commit 758c88b96963 ("realtek:
Whitespace and codestyle cleanup") causing rtl838x_read_phy() and
rtl838x_write_phy() to unconditionally return -ETIMEDOUT. As a result,
probing the device during boot fails:

    Error setting up netdev, freeing it again.
    rtl838x-eth: probe of 1b00a300.ethernet failed with error -5

Fix the bootloop caused by this regression with kernel 5.15 on rtl838x
devices, by properly returning 0 on success.

Tested on a Netgear GS108T v3, a Netgear GS310TP v1, a Zyxel GS1900-8HP
v1 and an HPE 1920-8G.

Fixes: 758c88b969639d0e6b684669d2e54dd1be3102f4 ("realtek: Whitespace and codestyle cleanup")
Tested-by: Stijn Segers <foss@volatilesystems.org>
Tested-by: Jan Hoffmann <jan@3e8.eu>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon commented:
@oliv3r: Please, for the love of god, split your commits into "whitespace changes *only*" and "changes in the code" next time, even if the latter look trivial at first glance. It has been a royal PITA to dig up which one of the 25 files contains the problematic change, and to track the actual line of code down within the changes applied to that file.

Borromini commented:
I will test this on my spare GS1900-8HP v1 in the next few days.

svanheule commented:
Sorry we missed this and thanks for the fix.

Could you update the coding style (of the four read/write functions)  to match rtl839x.c? That way `err` doesn't need to be initialized at the top, and the return value from `rtl838x_smi_wait_op()` isn't shadowed:

```c
    int err;
    ...
    err = rtl838x_smi_wait_op(100000);
    if (err)
        goto out;
    ...
out:
    ...
    return err;

oliv3r commented:
> @oliv3r: Please, for the love of god, split your commits into "whitespace changes _only_" and "changes in the code" next time, even if the latter look trivial at first glance. It has been a royal PITA to dig up which one of the 25 files contains the problematic change, and to actually track down the problematic changes within all the changes applied to that file within the same commit.

Sorry, but I do my best as I go. I can also say 'please, for the life of me, don't let such utter crap be committed', but here we are. Yes, it is hugely painful dealing with this, both ways.

Yeah, I admit, codestyle + whitespace should have been different commits, maybe. But then you can even split those out into each type of violation itself. Best is, not to commit violations :)

Borromini commented:
GS1900-8HP v1 boots as expected with the patch. I suppose you don't need me to open it up and boot a build without :innocent:.

janh commented:
I can confirm this fix works on HPE 1920-8G, although I wasn't seeing a bootloop here, it just failed to initialize the device:
```
[    2.736980] Error setting up netdev, freeing it again.
[    2.754199] rtl838x-eth: probe of 1b00a300.ethernet failed with error -5
```

hardfalcon commented:
@oliv3r: Sorry, my comment wasn't meant as harsh as it sounds now when I'm reading it again. I'm lucky y'all are contributing changes to this after all, because I wouldn't be able to write most of that code. :)

@svanheule: Sure, will do. I've noticed that the other files seem to use `errout:` instead of `out:`, so I'd use that label name as well if that's okay with you.
While I'm at it, should I also give the `rtl838x_pie_rule_write()` function the same treatment? That function currently always returns `0` even if the `rtl838x_write_pie_action()` call returns a non-zero value.

@Borromini: Thanks for the testing, much appreciated! :)

@janh: This might be due to me using the [linux-hardened patchset](https://github.com/anthraxx/linux-hardened). When bisecting this, sometimes the switch would hang with a backtrace but without rebooting, sometimes hang without a backtrace and without rebooting, sometimes the watchdog would kick in and reboot the device. I didn't test each of the different bisecting builds multiple times though, so tbh I can't really tell if the behaviour would randomly change for the same build. When I first encountered the problem, it was a bootloop, so that's the error description I kept using.

hardfalcon committed:
realtek: Improve error handling for rtl838x dsa driver on kernel 5.15

Make sure functions calling rtl838x_smi_wait_op() return its return
value in target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c.

Suggested-by: Sander Vanheule <sander@svanheule.net>
Signed-off-by: Pascal Ernster <git@hardfalcon.net>

hardfalcon committed:
realtek: Improve error handling in rtl838x_pie_rule_write() on kernel 5.15

In target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c,
make rtl838x_pie_rule_write() return non-zero value case of error.

Signed-off-by: Pascal Ernster <git@hardfalcon.net>

svanheule commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1345,24 +1345,27 @@ static int rtl838x_pie_rule_write(struct rtl838x_switch_priv *priv, int idx, str
-		goto err_out;
+		err = -1;
+		pr_err("Rule invalid\n");
+		goto errout;
```
You'll need to add braces around this block. It would also be better to provide a more meaningful error code like `EINVAL`.

hardfalcon replied:
Argh, this is really embarassing, I shouldn't have been writing that code so early in the morning… 🙈
Thanks, good catch! I've pushed a corrected version of the commit.

Borromini commented:
@svanheule Please do.

oliv3r commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1344,25 +1344,29 @@ static int rtl838x_pie_rule_write(struct rtl838x_switch_priv *priv, int idx, str
-	if (!pr->valid)
-		goto err_out;
+	if (!pr->valid) {
+		err = -1;
```
```suggestion
		err = -EINVAL;
```

oliv3r commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1335,7 +1335,7 @@ static int rtl838x_pie_rule_write(struct rtl838x_switch_priv *priv, int idx, str
 	struct table_reg *q = rtl_table_get(RTL8380_TBL_0, 1);
 	u32 r[18];
-	int err = 0;
+	int err;
```
```suggestion
	int err = -EINVAL;
```

best to keep some initialization here. Anything is better then nothing :)

hardfalcon replied:
@svanheule had suggested the exact opposite if I understood him correctly:
https://github.com/openwrt/openwrt/pull/11685#issuecomment-1370761187

oliv3r replied:
Looking at the entire function, we see that we guarantee assignment via https://github.com/openwrt/openwrt/blob/6783a8b8d872ca4507db987ed2e5e461fad9f291/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c#L1358 so then you can leave it uninitialized. Sorry for the noise :)

But always double check you don't end up with weird cases, you can return an uninitialized variable, that's why I personally prefer to always initialize things with known values (unless the first few lines immediately overwrites it).

hardfalcon replied:
No worries, I had asked myself the same question, but had assume that maybe modern compilers can throw a warning if a function returns an uninitialized variable (but I'm absolutetly no expert for C compilers).

oliv3r replied:
It can, and it does. But have you actually looked at the output of our kernel build? I'm sure we miss a few warnings there :D

And I only looked at the code snippet :p so I only saw your variable declaration change, + the first use, which was inside an if-guard :)

hardfalcon replied:
Thanks for the idea, this has actually made me try building with `CONFIG_WERROR=y` added to the kernel config, and fixing the stuff that's producing compiler warnings… 😄

oliv3r replied:
@hardfalcon Actually, I fixed almost all, if not all, of those warnings in my 170 large patch set. So 'in time' they should all drip in. I fixed them as part of the changes mostly, to not introduce too much ruckus :)

hardfalcon replied:
These are the additional changes I had to do to make it compile with `CONFIG_WERROR=y` about an hour ago. Most of it is simply adding a `__attribute__((unused))` here and there until the respective functions and variables are actually used, but there were also 2-3 minor code changes apart from that.

```diff
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
index 1fa92ae2..5f27d3a8 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/common.c
@@ -528,7 +528,7 @@ int rtl83xx_lag_del(struct dsa_switch *ds, int group, int port)
 }
 
 /* Allocate a 64 bit octet counter located in the LOG HW table */
-static int rtl83xx_octet_cntr_alloc(struct rtl838x_switch_priv *priv)
+static __attribute__((unused)) int rtl83xx_octet_cntr_alloc(struct rtl838x_switch_priv *priv)
 {
 	int idx;
 
@@ -1426,7 +1426,7 @@ static int rtl83xx_fib_event(struct notifier_block *this, unsigned long event, v
 			fib_info_hold(fib_work->fen_info.fi);
 
 		} else if (info->family == AF_INET6) {
-			struct fib6_entry_notifier_info *fen6_info = ptr;
+			struct fib6_entry_notifier_info __attribute__((unused)) *fen6_info = ptr;
 			pr_warn("%s: FIB_RULE ADD/DEL for IPv6 not supported\n", __func__);
 			kfree(fib_work);
 			return NOTIFY_DONE;
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
index 83a0441c..db147e0e 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/dsa.c
@@ -2013,7 +2013,7 @@ static bool rtl83xx_lag_can_offload(struct dsa_switch *ds,
 
 static int rtl83xx_port_lag_change(struct dsa_switch *ds, int port)
 {
-	struct rtl838x_switch_priv *priv = ds->priv;
+	struct rtl838x_switch_priv __attribute__((unused)) *priv = ds->priv;
 
 	pr_debug("%s: %d\n", __func__, port);
 	/* Nothing to be done... */
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
index 504b2982..35a06d14 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c
@@ -1291,7 +1291,7 @@ static void rtl838x_pie_rule_dump_raw(u32 r[])
 	pr_info("Sel    : %08x\n", r[17]);
 }
 
-static void rtl838x_pie_rule_dump(struct  pie_rule *pr)
+static __attribute__((unused)) void rtl838x_pie_rule_dump(struct  pie_rule *pr)
 {
 	pr_info("Drop: %d, fwd: %d, ovid: %d, ivid: %d, flt: %d, log: %d, rmk: %d, meter: %d tagst: %d, mir: %d, nopri: %d, cpupri: %d, otpid: %d, itpid: %d, shape: %d\n",
 		pr->drop, pr->fwd_sel, pr->ovid_sel, pr->ivid_sel, pr->flt_sel, pr->log_sel, pr->rmk_sel, pr->log_sel, pr->tagst_sel, pr->mir_sel, pr->nopri_sel,
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
index 7e4f13fb..b3fe5dbb 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl930x.c
@@ -990,7 +990,7 @@ static u32 rtl930x_l3_hash4(u32 ip, int algorithm, bool move_dip)
 	return hash;
 }
 
-static u32 rtl930x_l3_hash6(struct in6_addr *ip6, int algorithm, bool move_dip)
+static __attribute__((unused)) u32 rtl930x_l3_hash6(struct in6_addr *ip6, int algorithm, bool move_dip)
 {
 	u32 rows[16];
 	u32 hash;
@@ -1409,7 +1409,7 @@ static void rtl930x_get_l3_nexthop(int idx, u16 *dmac_id, u16 *interface)
 	*interface = v & 0x7f;
 }
 
-static int rtl930x_l3_mtu_del(struct rtl838x_switch_priv *priv, int mtu)
+static __attribute__((unused)) int rtl930x_l3_mtu_del(struct rtl838x_switch_priv *priv, int mtu)
 {
 	int i;
 
@@ -1458,7 +1458,7 @@ static int rtl930x_l3_mtu_add(struct rtl838x_switch_priv *priv, int mtu)
 }
 
 /* Creates an interface for a route by setting up the HW tables in the SoC */
-static int rtl930x_l3_intf_add(struct rtl838x_switch_priv *priv, struct rtl838x_l3_intf *intf)
+static __attribute__((unused)) int rtl930x_l3_intf_add(struct rtl838x_switch_priv *priv, struct rtl838x_l3_intf *intf)
 {
 	int i, intf_id, mtu_id;
 	/* number of MTU-values < 16384 */
@@ -1681,7 +1681,7 @@ static void rtl930x_write_pie_templated(u32 r[], struct pie_rule *pr, enum templ
 	}
 }
 
-static void rtl930x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
+static __attribute__((unused)) void rtl930x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
 {
 	pr->stacking_port = r[6] & BIT(31);
 	pr->spn = (r[6] >> 24) & 0x7f;
diff --git a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
index 92f06622..f67545ad 100644
--- a/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
+++ b/target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl931x.c
@@ -1113,7 +1113,7 @@ static void rtl931x_write_pie_templated(u32 r[], struct pie_rule *pr, enum templ
 	}
 }
 
-static void rtl931x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
+static __attribute__((unused)) void rtl931x_read_pie_fixed_fields(u32 r[], struct pie_rule *pr)
 {
 	pr->mgnt_vlan = r[7] & BIT(31);
 	if (pr->phase == PHASE_IACL)
diff --git a/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c b/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
index 857d389f..c0447ac4 100644
--- a/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
+++ b/target/linux/realtek/files-5.15/drivers/net/ethernet/rtl838x_eth.c
@@ -161,7 +161,7 @@ static void rtl931x_create_tx_header(struct p_hdr *h, unsigned int dest_port, in
 		h->cpu_tag[2] = (BIT(5) | (prio & 0x1f)) << 8;
 }
 
-static void rtl93xx_header_vlan_set(struct p_hdr *h, int vlan)
+static __attribute__((unused)) void rtl93xx_header_vlan_set(struct p_hdr *h, int vlan)
 {
 	h->cpu_tag[2] |= BIT(4); /* Enable VLAN forwarding offload */
 	h->cpu_tag[2] |= (vlan >> 8) & 0xf;
@@ -862,7 +862,7 @@ static void rtl839x_setup_notify_ring_buffer(struct rtl838x_eth_priv *priv)
 
 	/* Setup notification events */
 	sw_w32_mask(0, 1 << 14, RTL839X_L2_CTRL_0); /* RTL8390_L2_CTRL_0_FLUSH_NOTIFY_EN */
-	sw_w32_mask(0, 1 << 12, RTL839X_L2_NOTIFICATION_CTRL); /* SUSPEND_NOTIFICATION_EN
+	sw_w32_mask(0, 1 << 12, RTL839X_L2_NOTIFICATION_CTRL); /* SUSPEND_NOTIFICATION_EN */
 
 	/* Enable Notification */
 	sw_w32_mask(0, 1 << 0, RTL839X_L2_NOTIFICATION_CTRL);
@@ -1957,7 +1957,7 @@ static int rtl930x_mdio_reset(struct mii_bus *bus)
 			break;			/* Serdes: Value = 0 */
 		case PHY_INTERFACE_MODE_HSGMII:
 			private_poll_mask |= BIT(i);
-			/* fallthrough */
+			fallthrough;
 		case PHY_INTERFACE_MODE_USXGMII:
 			v |= BIT(mac_type_bit[i]);
 			uses_usxgmii = true;
diff --git a/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c b/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
index 77332527..cd513cc7 100644
--- a/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
+++ b/target/linux/realtek/files-5.15/drivers/net/phy/rtl83xx-phy.c
@@ -744,7 +744,7 @@ static int rtl8380_configure_int_rtl8218b(struct phy_device *phydev)
 	rtl8218b_6276B_hwEsd_perport = (void *)h + sizeof(struct fw_header) + h->parts[9].start;
 
 	if (sw_r32(RTL838X_DMY_REG31) == 0x1) {
-		int ipd_flag = 1;
+		int __attribute__((unused)) ipd_flag = 1;
 	}
 
 	val = phy_read(phydev, 0);
@@ -3330,7 +3330,7 @@ static void rtl931x_sds_rx_rst(u32 sds)
 	mdelay(50);
 }
 
-static void rtl931x_sds_disable(u32 sds)
+static __attribute__((unused)) void rtl931x_sds_disable(u32 sds)
 {
 	u32 v = 0x1f;
 
diff --git a/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch b/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
index 8109abd2..213fe8a2 100644
--- a/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
+++ b/target/linux/realtek/patches-5.15/704-include-linux-add-phy-hsgmii-mode.patch
@@ -31,3 +31,18 @@ Submitted-by: Birger Koblitz <git@birger-koblitz.de>
  	case PHY_INTERFACE_MODE_QSGMII:
  		return "qsgmii";
  	case PHY_INTERFACE_MODE_TRGMII:
+diff --git a/drivers/net/phy/phylink.c b/drivers/net/phy/phylink.c
+index afe69295..4da98bc7 100644
+--- a/drivers/net/phy/phylink.c
++++ b/drivers/net/phy/phylink.c
+@@ -362,6 +362,10 @@ void phylink_get_linkmodes(unsigned long *linkmodes, phy_interface_t interface,
+ 		caps |= MAC_10000FD | MAC_5000FD | MAC_2500FD;
+ 		fallthrough;
+ 
++	case PHY_INTERFACE_MODE_HSGMII:
++		caps |= MAC_2500FD;
++		fallthrough;
++
+ 	case PHY_INTERFACE_MODE_RGMII_TXID:
+ 	case PHY_INTERFACE_MODE_RGMII_RXID:
+ 	case PHY_INTERFACE_MODE_RGMII_ID:
diff --git a/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch b/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
index a80482d9..5c286b15 100644
--- a/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
+++ b/target/linux/realtek/patches-5.15/711-net-phy-add-an-MDIO-SMBus-library.patch
@@ -131,12 +131,12 @@ Signed-off-by: Antoine Tenart <antoine.tenart@bootlin.com>
 + * specified to be present in SFP modules.  These correspond with PHY
 + * addresses 16 and 17.  Disallow access to these "phy" addresses.
 + */
-+static bool i2c_mii_valid_phy_id(int phy_id)
++static __attribute__((unused)) bool i2c_mii_valid_phy_id(int phy_id)
 +{
 +	return phy_id != 0x10 && phy_id != 0x11;
 +}
 +
-+static unsigned int i2c_mii_phy_addr(int phy_id)
++static __attribute__((unused)) unsigned int i2c_mii_phy_addr(int phy_id)
 +{
 +	return phy_id + 0x40;
 +}
```

oliv3r commented on target/linux/realtek/files-5.15/drivers/net/dsa/rtl83xx/rtl838x.c:
```diff
@@ -1798,7 +1802,7 @@ int rtl838x_smi_wait_op(int timeout)
 int rtl838x_read_phy(u32 port, u32 page, u32 reg, u32 *val)
 {
-	int err = -ETIMEDOUT;
+	int err;
```
same here, keep something as default.

oliv3r reviewed (commented):
overall good improvements! thanks!

hardfalcon commented:
> @svanheule I'm fine with a Tested-by tag.

Added.

svanheule commented:
@hardfalcon Thanks for the updated. Perhaps more of a cosmetic thing, but I think the commit message for the fix patch could be improved to explain what's causing the issue. 

Something like the following perhaps?
```
realtek: return correct error value for phy ops

A behavioural change was introduced with commit 758c88b96963 ("realtek:
Whitespace and codestyle cleanup") causing rtl838x_read_phy() and
rtl838x_write_phy() to unconditionally return -ETIMEDOUT. As a result,
probing the device during boot fails:

    Error setting up netdev, freeing it again.
    rtl838x-eth: probe of 1b00a300.ethernet failed with error -5

Fix the bootloop caused by this regression with kernel 5.15 on rtl838x
devices, by properly returning 0 on success.
```

hardfalcon commented:
@svanheule: Done, thanks for the suggestion! :)

@Spudz76: Thanks a lot, gotta admit I wasn't aware this option exists 🤯

svanheule commented:
Thanks! Merged with commits:

- de2dc3feaea7 realtek: return correct error value for phy ops
- a188536ef650 realtek: 5.15: Improve rtl838x dsa driver error handling
- 720b2431716b realtek: 5.15: Improve error handling in rtl838x_pie_rule_write()
